gdal/frmts/wms/gdalwmscache.cpp

```diff
@@ -31,89 +33,269 @@
+        // Warns if it fails to write, but returns success
+        CPLString soFilePath = GetFilePath( pszKey );
+        MakeDirs( CPLGetDirname(soFilePath) );
+        if ( CPLCopyFile( soFilePath, osFileName ) == CE_None)
+            return CE_None;
+        // Warn if it fails after folder creation
+        CPLError( CE_Warning, CPLE_FileIO, "Error writing to WMS cache %s",
+                 m_soPath.c_str() );
+        return CE_None;
+    }
+
+    virtual enum GDALWMSCacheItemStatus GetItemStatus(const char *pszKey) const CPL_OVERRIDE
+    {
+        VSIStatBufL  sStatBuf;
+        if( VSIStatL( GetFilePath(pszKey), &sStatBuf ) == 0 )
         {
-            m_cache_path = usercache_path;
+            long seconds = static_cast<long>( time( nullptr ) - sStatBuf.st_mtime );
+            return seconds < m_nExpires ? CACHE_ITEM_OK : CACHE_ITEM_EXPIRED;
         }
-        else
+        return  CACHE_ITEM_NOT_FOUND;
+    }
+
+    virtual GDALDataset* GetDataset(const char *pszKey, char **papszOpenOptions) const CPL_OVERRIDE
+    {
+        return reinterpret_cast<GDALDataset*>(
+                    GDALOpenEx( GetFilePath( pszKey ), GDAL_OF_RASTER |
+                               GDAL_OF_READONLY | GDAL_OF_VERBOSE_ERROR, nullptr,
+                               papszOpenOptions, nullptr ) );
+    }
+
+    virtual void Clean() CPL_OVERRIDE
+    {
+        char **papszList = VSIReadDirRecursive( m_soPath );
+        if( papszList == nullptr )
         {
-            m_cache_path = "./gdalwmscache";
+            return;
         }
+
+        int counter = 0;
+        std::vector<int> toDelete;
+        long nSize = 0;
+        time_t nTime = time( nullptr );
+        while( papszList[counter] != nullptr )
+        {
+            const char* pszPath = CPLFormFilename( m_soPath, papszList[counter], nullptr );
+            VSIStatBufL sStatBuf;
+            if( VSIStatL( pszPath, &sStatBuf ) == 0 )
+            {
+                if( !VSI_ISDIR( sStatBuf.st_mode ) )
+                {
+                    long seconds = static_cast<long>( nTime - sStatBuf.st_mtime );
+                    if(seconds > m_nExpires)
+                    {
+                        toDelete.push_back(counter);
+                    }
+
+                    nSize += static_cast<long>( sStatBuf.st_size );
+                }
+            }
+            counter++;
+        }
+
+        if( nSize > m_nMaxSize )
+        {
+            CPLDebug( "WMS", "Delete %u items from cache", static_cast<unsigned int>(toDelete.size()) );
+            for( size_t i = 0; i < toDelete.size(); ++i )
+            {
+                const char* pszPath = CPLFormFilename( m_soPath,
+                                                       papszList[toDelete[i]],
+                                                       nullptr );
+                VSIUnlink( pszPath );
+            }
+        }
+
+        CSLDestroy(papszList);
+
+        // Prevent very frequently execution
+        CPLSleep(15);
```

Comment:
This could delay the termination of the thread when joining it. Perhaps add a time based logic at the place where the thread is started ?

Reply:
Fixed